### DISCUSSION TITLE
Backend usage with validations

### DISCUSSION THREAD

**Discussion by gerwinbrunner** [statelyai/xstate#3508] [General]
Hi,

I want to use xState in the backend to build a state based workflow process.

We have "Cases" can can go from `DRAFT` to `OPEN` to `CLOSE`

Before they can go from `DRAFT` to `OPEN` there is a long list of validation criteria that need to be met.
i.E: "title is present", "lossamount is present", "translation is present", "approved by at least 2 people" , ...

What I want is that then the transition is triggered the validations are run and then:
A) all validations pass and the transition to OPEN goes thru
or
B) some validations fail and we stay in DRAFT - I also get a list of validation errors to present to the user

I tried to implement this by using guards but it does not really give me the possibility to return the errors.

How would I do this? 
Any hints?

**Comment by scarabaeus** [comment]
It should work on the backend the same as it does on the frontend -- you'll need to set the errors in context, using an action, when the appropriate guard evaluates to `true`.  Check out the accepted answer on this question for an example: https://github.com/statelyai/xstate/discussions/1969

_Paraphrased from linked answer:_
```
{
  on: {
    TRANSITION_TO_OPEN: [
      {
        cond: "isValueMissing",
        actions: ["setValue", "setValueMissingError"],
      },
      {
        cond: "isValueInTheFuture",
        actions: ["setValue", "setValueInTheFutureError"],
      },
      {
        actions: ["setValue", "removeError"],
      },
    ],
  },
}
```

**Reply by gerwinbrunner** [reply]
@scarabaeus Thanks for your answer.
I have already tried this approach but it has two major issues.

A) Adds a lot of additional code
Let's assume the `isValueMissing` function could return up to 5 different errors that could happen and I want to pass those errors to the frontend.  
It is not really possible to split the function in multiple guards - one per error 


Do you have any other idea how I could set /add the errors to errors list from a condition?
So the ideal way would be to add the errors directly in the condition in the `false` case.
